src/librustc_codegen_ssa/back/linker.rs

```diff
@@ -538,6 +539,21 @@ impl<'a> Linker for GccLinker<'a> {
             if let Err(e) = res {
                 self.sess.fatal(&format!("failed to write lib.def file: {}", e));
             }
+        } else if is_windows {
+            let res: io::Result<()> = try {
+                let mut f = BufWriter::new(File::create(&path)?);
+
+                // .def file similar to MSVC one but without LIBRARY section
+                // because LD doesn't like when it's empty
+                writeln!(f, "EXPORTS")?;
+                for symbol in self.info.exports[&crate_type].iter() {
+                    debug!("  _{}", symbol);
+                    writeln!(f, "  {}", symbol)?;
+                }
+            };
+            if let Err(e) = res {
+                self.sess.fatal(&format!("failed to write list.def file: {}", e));
+            }
```

Comment:
I forgot to ask - what exactly makes `.def` files work with `ld` in this PR?
The `--version-script` -> `list.def` change caused test failures in https://github.com/rust-lang/rust/pull/70852 despite the export list being the same.

Reply:
> I forgot to ask - what exactly makes .def files work with ld in this PR?

TL;DR we now manually add dllexport/dllimport.

With version-scripts LD sort of scans the symbols to decide when to automatically add dllexport.
With `.def` it doesn't use automatic dllexport but dllexport can be manually enabled by what this PR does or writing `DATA` next to the symbol on the list.

Previously most of the tests worked before because of automatic dllimport, broken tests were using statics which don't work with automatic dllimport.